src/anomalib/pre_processing/utils/transform.py

```diff
@@ -0,0 +1,182 @@
+        Dictionary mapping stages to their transforms.
+    """
+    transforms: dict[str, Transform] = {}
+    stage_lookup = {
+        "fit": "train",
+        "validate": "val",
+        "test": "test",
+        "predict": "test",
+    }
+
+    for dataloader in dataloaders:
+        if not hasattr(dataloader, "dataset") or not hasattr(dataloader.dataset, "transform"):
+            continue
+
+        for stage in stage_lookup:
+            if hasattr(dataloader, f"{stage}_dataloader"):
+                transforms[stage_lookup[stage]] = dataloader.dataset.transform
+
+    return transforms
+
+
+def set_datamodule_transforms(datamodule: AnomalibDataModule, transforms: dict[str, Transform | None]) -> None:
+    """Set transforms to a datamodule.
+
+    Args:
+        datamodule: The datamodule to propagate transforms to.
+        transforms: Dictionary mapping stages to their transforms.
+    """
+    for stage in ["fit", "validate", "test", "predict"]:
+        transform = get_stage_transform(stage, transforms)
+        if transform is not None:
+            set_datamodule_stage_transform(datamodule, transform, stage)
+
+
+def set_dataloaders_transforms(dataloaders: Sequence[DataLoader], transforms: dict[str, Transform | None]) -> None:
+    """Set transforms to dataloaders.
+
+    Args:
+        dataloaders: The dataloaders to propagate transforms to.
+        transforms: Dictionary mapping stages to their transforms.
+    """
+    stage_mapping = {
+        "fit": "train",
+        "validate": "val",
+        "test": "test",
+        "predict": "test",  # predict uses test transform
+    }
+
+    for loader in dataloaders:
+        if not hasattr(loader, "dataset"):
+            continue
+
+        for stage in stage_mapping:
+            if hasattr(loader, f"{stage}_dataloader"):
+                transform = transforms.get(stage_mapping[stage])
+                if transform is not None:
+                    set_dataloader_transform([loader], transform)
+
+
+def set_datamodule_stage_transform(datamodule: AnomalibDataModule, transform: Transform, stage: str) -> None:
+    """Set a transform for a specific stage in a AnomalibDataModule.
+
+    Args:
+        datamodule: The AnomalibDataModule to set the transform for.
+        transform: The transform to set.
+        stage: The stage to set the transform for.
+
+    Note:
+        The stage parameter maps to dataset attributes as follows:
+        - 'fit' -> 'train_data'
+        - 'validate' -> 'val_data'
+        - 'test' -> 'test_data'
+        - 'predict' -> 'test_data'
+    """
+    stage_datasets = {
+        "fit": "train_data",
+        "validate": "val_data",
+        "test": "test_data",
+        "predict": "test_data",
+    }
+
+    dataset_attr = stage_datasets.get(stage)
+    if dataset_attr and hasattr(datamodule, dataset_attr):
+        dataset = getattr(datamodule, dataset_attr)
+        if hasattr(dataset, "transform"):
+            dataset.transform = transform
```

Comment:
Would there be a way to assign the transforms to the datamodule before the datasets are instantiated, instead of overwriting them here? That way the datasets would always have the right transform, which would be less prone to bugs.

I guess it is done this way because `setup` callback hook gets called after the datamodule's `setup` hook, so by the time that we set up the pre-processor, the datasets are already created. I just wonder if there is a different way to do it which does not involve overwriting the transforms.

Reply:
In this PR https://github.com/openvinotoolkit/anomalib/pull/2239, the setup logic completely changes. It might be an idea to visit it when working on this PR instead of addressing here